### PR TITLE
Add Codec Methods

### PR DESCRIPTION
**Pull request by Yip-Jia-Qi**
## What?

Added the ability to use a codec encoders and decoders for speech separation
Registered new encoders, decoders and separator under the separation recipe

## Why?

This allows a new way of perform speech separation using speech representations of neural audio codecs

## See also

This is based on the paper https://arxiv.org/abs/2406.12434 and is part of ESPnet-Codec

### CODE REVIEW

**Comment by ftshijt** [comment]
Many thanks! @Emrys365 May I kindly request your help in reviewing the PR~

@Yip-Jia-Qi could you please also fix the CI test issues?

**Comment by Emrys365** [comment]
> Many thanks! @Emrys365 May I kindly request your help in reviewing the PR~

Sure. This is a very cool PR！

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+optim: adam
+init: xavier_uniform
+max_epoch: 1
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:3]: Is this for debugging? If so, it should be removed.

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    loss_type: si_snr
+encoder: codec
+encoder_conf:
+    channel: 512
+    codec_choice: "ESPnet"
+    codec_fs: 16000 #must match the codec
+    sample_fs: 8000
+    device: "cuda:0"
+    dump_audio: True
+    checkpoint_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/train.total_count.best.pth"
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:39]: Avoid using absolute paths that are not reusable by other users. Please also add information on how the checkpoint should be prepared/obtained.

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    config_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/config.yaml"
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:40]: ditto

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+decoder: codec
+decoder_conf:
+    channel: 512
+    codec_choice: "ESPnet"
+    codec_fs: 16000 #must match the codec
+    sample_fs: 8000
+    device: "cuda:0"
+    dump_audio: True
+    checkpoint_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/train.total_count.best.pth"
+    config_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/config.yaml"
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:50]: ditto

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    patience: 1
+model_conf:
+    loss_type: si_snr
+encoder: codec
+encoder_conf:
+    channel: 512
+    codec_choice: "ESPnet"
+    codec_fs: 16000 #must match the codec
+    sample_fs: 8000
+    device: "cuda:0"
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:37]: This will impede the use of multiple GPUs I think.

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    dump_audio: True
+    checkpoint_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/train.total_count.best.pth"
+    config_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/config.yaml"
+decoder: codec
+decoder_conf:
+    channel: 512
+    codec_choice: "ESPnet"
+    codec_fs: 16000 #must match the codec
+    sample_fs: 8000
+    device: "cuda:0"
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:47]: ditto

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    dump_audio: True
+    checkpoint_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/train.total_count.best.pth"
+    config_path: "/home/jiaqi006/code/new_espnet/soundstream_model/16khz_soundstream/config.yaml"
+separator: codec
+separator_conf:
+    # input_dim: 512 #not needed because enh.py infers this.
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:53]:
```suggestion
```

File: egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    - loss
+    - min
+keep_nbest_models: 1
+scheduler: reducelronplateau
+scheduler_conf:
+    mode: min
+    factor: 0.5
+    patience: 1
+model_conf:
+    loss_type: si_snr
```
Comment (Emrys365) [inline egs2/librimix/enh1/conf/tuning/train_enh_codecformer.yaml:30]: This argument has been deprecated. Please follow https://github.com/espnet/espnet/blob/master/egs2/wsj0_2mix/enh1/conf/tuning/train_enh_conv_tasnet.yaml#L52-L60 to set the criterion.

File: espnet2/enh/decoder/codec_decoder.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        channel: int,  # this is the expected output size of the continuous codec output
+        codec_choice: str,
+        codec_fs: int,
+        sample_fs: int,
+        device: str = "cpu",
+        dump_audio: bool = False,
+        checkpoint_path: str = None,
+        config_path: str = None,
+        max_token_per_frame: int = 32,
+    ):
```
Comment (Emrys365) [inline espnet2/enh/decoder/codec_decoder.py:23]: It is better to add a docstring to explain each argument here. For example, what is the difference between `codec_fs` and `sample_fs`, and how can we prepare a customized `checkpoint_path` and `config_path` (what should these files look like)?

File: espnet2/enh/decoder/codec_decoder.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        for param in self.codec.parameters():
+            param.requires_grad = False
+        self._output_dim = channel
+
+    def forward(self, input: torch.Tensor, ilens: torch.Tensor, fs: int = None):
+        """Forward.
+
+        Args:
+            input (torch.Tensor): spectrum [Batch, T, F]
```
Comment (Emrys365) [inline espnet2/enh/decoder/codec_decoder.py:46]: It is better to consider a future extension to support multi-channel input speech at this stage. Currently the interface always assume the input to be single-channel.

File: espnet2/enh/decoder/codec_decoder.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+        # T might have changed due to model. If so, fix it here
+        T_origin = ilens.max()
+        if wav.shape[-1] != T_origin:
+
+            T_est = wav.shape[-1]
+            if T_origin > T_est:
+                wav = F.pad(wav, (0, T_origin - T_est))
+            else:
+                wav = wav[:, :T_origin]
```
Comment (Emrys365) [inline espnet2/enh/decoder/codec_decoder.py:66]: You can simply replace these lines with
```python
wav = F.pad(wav, (0, T_origin - wav.shape[-1]))
```

`F.pad` can automatically do padding or truncation according to the specified padding value.

File: espnet2/enh/encoder/codec_encoder.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        channel: int,  # this is the expected output size of the continuous codec output
+        codec_choice: str,
+        codec_fs: int,
+        sample_fs: int,
+        device: str = "cpu",
+        dump_audio: bool = False,
+        checkpoint_path: str = None,
+        config_path: str = None,
+        max_token_per_frame: int = 32,
+    ):
```
Comment (Emrys365) [inline espnet2/enh/encoder/codec_encoder.py:22]: ditto

File: espnet2/enh/encoder/codec_encoder.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            dump_audio=dump_audio,
+            checkpoint_path=checkpoint_path,
+            config_path=config_path,
+        )
+
+        self.codec_fs = codec_fs
+        self.sample_fs = sample_fs
+        if codec_fs != sample_fs:
+            self.dac_sampler = T.Resample(sample_fs, codec_fs).to(device)
+        # self.org_sampler = T.Resample(DAC_sample_rate, input_sample_rate)
```
Comment (Emrys365) [inline espnet2/enh/encoder/codec_encoder.py:37]: Remove this line.

File: espnet2/enh/decoder/codec_decoder.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        device = x.device
+
+        # Implement some checks on the input
+        assert len(x.shape) == 3
+        B, C, T = x.shape
+        assert C == 1  # model should only be handling single channel
+
+        # Resamples the audio from the input rate to the dac model's rate
+
+        x_resamp = self.dac_sampler(x)
```
Comment (Emrys365) [inline espnet2/enh/decoder/codec_decoder.py:84]: Why do you have to go to cpu for resampling? I can't understand this.

BTW, you can use torchaudio.functional.resample to replace the current T.Resample because the former can automatically use the same device as its input.

File: espnet2/enh/decoder/seanet_decoder.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        encdec_dilation_base: int = 2,
+        encdec_causal: bool = False,
+        encdec_pad_mode: str = "reflect",
+        encdec_true_skip: bool = False,
+        encdec_compress: int = 2,
+        encdec_lstm: int = 2,
+        decoder_trim_right_ratio: float = 1.0,
+        decoder_final_activation: Optional[str] = None,
+        decoder_final_activation_params: Optional[dict] = None,
+    ):
```
Comment (Emrys365) [inline espnet2/enh/decoder/seanet_decoder.py:39]: Please add a docstring to explain each argument.

File: espnet2/enh/decoder/seanet_decoder.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+        # T might have changed due to model. If so, fix it here
+        T_origin = ilens.max()
+        if wav.shape[-1] != T_origin:
+
+            T_est = wav.shape[-1]
+            if T_origin > T_est:
+                wav = F.pad(wav, (0, T_origin - T_est))
+            else:
+                wav = wav[:, :T_origin]
```
Comment (Emrys365) [inline espnet2/enh/decoder/seanet_decoder.py:95]: You can simply replace these lines with
```python
wav = F.pad(wav, (0, T_origin - wav.shape[-1]))
```

`F.pad` can automatically do padding or truncation according to the specified padding value.

File: espnet2/enh/encoder/seanet_encoder.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        encdec_kernel_size: int = 7,
+        encdec_residual_kernel_size: int = 7,
+        encdec_last_kernel_size: int = 7,
+        encdec_dilation_base: int = 2,
+        encdec_causal: bool = False,
+        encdec_pad_mode: str = "reflect",
+        encdec_true_skip: bool = False,
+        encdec_compress: int = 2,
+        encdec_lstm: int = 2,
+    ):
```
Comment (Emrys365) [inline espnet2/enh/encoder/seanet_encoder.py:37]: ditto

File: espnet2/enh/layers/sb_transformer_block.py [hunk 18]
```diff
@@ -0,0 +1,876 @@
...
+Modifications made by Jia Qi Yip 2024
+"""
+
+import copy
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# from ..utils.Transformer import PositionalEncoding, TransformerEncoder
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:21]: Remove this line if it is not used.

File: espnet2/enh/layers/sb_transformer_block.py [hunk 19]
```diff
@@ -0,0 +1,876 @@
...
+        input_shape=None,
+        kdim=None,
+        vdim=None,
+        dropout=0.1,
+        activation="relu",
+        use_positional_encoding=False,
+        norm_before=False,
+        attention_type="regularMHA",
+    ):
+        super(SBTransformerBlock, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:76]:
```suggestion
        super().__init__()
```

File: espnet2/enh/layers/sb_transformer_block.py [hunk 20]
```diff
@@ -0,0 +1,876 @@
...
+
+    Example
+    -------
+    >>> x = torch.randn(5, 10, 20)
+    >>> GLN = GlobalLayerNorm(10, 3)
+    >>> x_norm = GLN(x)
+    """
+
+    def __init__(self, dim, shape, eps=1e-8, elementwise_affine=True):
+        super(GlobalLayerNorm, self).__init__()
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:226]:
```suggestion
        super().__init__()
```

File: espnet2/enh/layers/sb_transformer_block.py [hunk 21]
```diff
@@ -0,0 +1,876 @@
...
+
+    Example
+    -------
+    >>> x = torch.randn(5, 10, 20)
+    >>> CLN = CumulativeLayerNorm(10)
+    >>> x_norm = CLN(x)
+    """
+
+    def __init__(self, dim, elementwise_affine=True, eps=1e-8):
+        super(CumulativeLayerNorm, self).__init__(
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:290]:
```suggestion
        super().__init__(
```

File: espnet2/enh/layers/sb_transformer_block.py [hunk 22]
```diff
@@ -0,0 +1,876 @@
...
+
+Modifications made by Jia Qi Yip 2024
+"""
+
+import math
+from typing import Optional
+
+import numpy as np
+import torch
+import torch.nn as nn
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:333]: Please put all imports on the top of the file.

File: espnet2/enh/layers/sb_transformer_block.py [hunk 23]
```diff
@@ -0,0 +1,876 @@
...
+                d_model=d_model,
+                dropout=dropout,
+                kdim=kdim,
+                vdim=vdim,
+            )
+
+        # elif attention_type == "RelPosMHAXL":
+        #     self.self_att = sb.nnet.attention.RelPosMHAXL(
+        #         d_model, nhead, dropout, mask_pos_future=causal
+        #     )
```
Comment (Emrys365) [inline espnet2/enh/layers/sb_transformer_block.py:446]: Remove unused lines.

File: espnet2/enh/separator/codec_separator.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+import torch
+import torch.nn as nn
+from packaging.version import parse as V
+from torch.nn.utils import weight_norm
+
+from espnet2.enh.layers.sb_transformer_block import SBTransformerBlock
+from espnet2.enh.layers.snake1d import Snake1d
+from espnet2.enh.separator.abs_separator import AbsSeparator
+
+is_torch_1_9_plus = V(torch.__version__) >= V("1.9.0")
```
Comment (Emrys365) [inline espnet2/enh/separator/codec_separator.py:13]: Remove this line since it is not referenced.

File: espnet2/enh/separator/codec_separator.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        activation: str = "ELU",
+        activation_params: dict = {"alpha": 1.0},
+        trf_num_layers: int = 16,
+        trf_num_heads: int = 8,
+        trf_feedforward: int = 2048,
+        trf_dropout: float = 0.1,
+        trf_use_positional_encoding: bool = True,
+        trf_norm_before: bool = True,
+    ):
+        """Dual-Path RNN (DPRNN) Separator
```
Comment (Emrys365) [inline espnet2/enh/separator/codec_separator.py:33]: The name should be modified. Add a reference paper if possible.

File: espnet2/enh/separator/codec_separator.py [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+
+        Args:
+            input_dim: input feature dimension
+            num_spk: number of speakers
+            activation (str): Activation function.
+            activation_params (dict): Parameters to provide to the activation function
```
Comment (Emrys365) [inline espnet2/enh/separator/codec_separator.py:39]: Please explain all arguments in the docstring.

File: espnet2/enh/separator/codec_separator.py [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+
+        ##OLD VERSION
+        # masks = self.masker(x)
+        # #[B,N*num_outputs,T]
+        # masks = masks.view(B, N, self.num_outputs,T).permute(0,1,3,2)
+        # masks = [
+        #     self.activation(self.output(masks[...,i]) * self.output_gate(masks[...,i])).permute(0,2,1)
+        #     for i in range(self.num_outputs)
+        # ]
+
```
Comment (Emrys365) [inline espnet2/enh/separator/codec_separator.py:139]: Remove unused lines.

File: espnet2/tasks/enh.py [hunk 28]
```diff
@@ -130,7 +142,13 @@
 
 decoder_choices = ClassChoices(
     name="decoder",
-    classes=dict(stft=STFTDecoder, conv=ConvDecoder, same=NullDecoder),
+    classes=dict(
+        stft=STFTDecoder,
+        conv=ConvDecoder,
+        same=NullDecoder,
+        codec=CodecDecoder,
+        seanet=SEANetEnhDecoder,
```
Comment (Emrys365) [inline espnet2/tasks/enh.py:150]: Can you add an additional yaml file under `egs2/librimix/enh1/conf/tuning` for the `seanet` separator? This helps clarify how we can use the added model.

**Review by Emrys365 — COMMENTED** [review]
Thanks a lot for your contribution @Yip-Jia-Qi! I have left several general comments.

**Comment by Emrys365** [comment]
In addition, could you add unit tests for the added models under `test/espnet2/enh`?

> Generally, for a new model, we would add tests in `test/espnet2/enh/test_espnet_model.py`, `test/espnet2/enh/encoder/test_[new encoder name].py`, `test/espnet2/enh/decoder/test_[new decoder name].py`, and `test/espnet2/enh/separator/test_[new model name].py`.
